### ISSUE TITLE
Phase 57.3: LoadBalancer — Auxiliary Losses & Expert Capacity Management

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#1084]
### Overview
The **LoadBalancer** ensures even expert utilization across all MoE layers, preventing routing collapse (where all tokens route to a few experts) and expert starvation (where some experts receive no tokens). It implements auxiliary balancing losses, capacity management strategies, and adaptive load redistribution to maintain healthy expert utilization throughout training.

### Architecture

```typescript
interface LoadBalancerConfig {
  balancingStrategy: 'auxiliary_loss' | 'expert_choice' | 'sinkhorn' | 'adaptive';
  auxiliaryLossWeight: number;       // λ_aux — weight for balancing loss (typically 0.01)
  routerZLossWeight: number;         // λ_z — weight for z-loss (typically 0.001)
  importanceLossWeight: number;      // λ_imp — importance-based loss weight
  capacityFactor: number;            // C — expert capacity = (N/E) * C (typically 1.0-1.5)
  tokenDropThreshold: number;        // Max fraction of tokens that can be dropped
  minExpertUtilization: number;      // Minimum fraction of tokens per expert
  smoothingWindow: number;           // EMA window for load statistics
  rebalanceInterval: number;         // Steps between forced rebalancing
  sinkHornIterations: number;        // Iterations for Sinkhorn-Knopp balancing
}

interface LoadBalanceMetrics {
  expertUtilization: number[];       // Per-expert utilization ratios
  coefficientOfVariation: number;    // CV of expert loads (lower = more balanced)
  maxLoadImbalance: number;          // max(load) / mean(load)
  tokenDropRate: number;             // Fraction of tokens dropped
  auxiliaryLoss: number;             // Current auxiliary loss value
  routerZLoss: number;               // Current z-loss value
  expertCollapse: boolean;           // Whether routing collapse detected
  deadExperts: number[];             // Indices of experts receiving <1% tokens
  overloadedExperts: number[];       // Indices exceeding capacity
  balanceScore: number;              // 0-1 overall balance score
}

class LoadBalancer {
  // Loss computation
  computeAuxiliaryLoss(routerLogits: Tensor, expertMask: Tensor): number;
  computeSwitchBalancingLoss(routerProbs: Tensor, expertAssignments: Tensor): number;
  computeImportanceLoss(gateValues: Tensor): number;
  computeRouterZLoss(routerLogits: Tensor): number;
  computeCombinedLoss(): CombinedLossResult;
  
  // Capacity management
  enforceExpertCapacity(assignments: Tensor, capacity: number): Tensor;
  computeOptimalCapacityFactor(currentLoads: number[]): number;
  adaptCapacityFactor(metrics: LoadBalanceMetrics): number;
  
  // Token dropping
  selectTokensToDrop(overflowTokens: Tensor, policy: string): Tensor;
  rerouteDroppedTokens(tokens: Tensor, originalRouting: RoutingDecision): RoutingDecision;
  handleResidualConnection(droppedTokens: Tensor): Tensor;
  
  // Expert health monitoring
  detectRoutingCollapse(history: LoadBalanceMetrics[]): boolean;
  identifyDeadExperts(utilization: number[], threshold: number): number[];
  reviveDeadExperts(deadIndices: number[]): void;
  computeExpertEntropy(routerProbs: Tensor): number;
  
  // Adaptive balancing
  sinkHornBalancing(scores: Tensor, iterations: number): Tensor;
  adjustRoutingTemperature(metrics: LoadBalanceMetrics): number;
  scheduleCapacityAnnealing(startCF: number, endCF: number, steps: number): Schedule;
  
  // Monitoring
  getLoadStatistics(): LoadBalanceMetrics;
  trackLoadHistory(windowSize: number): LoadBalanceMetrics[];
  generateBalanceReport(): BalanceReport;
}
```

#### Switch Balancing Loss (Fedus et al.)
```
L_aux = α · N · Σ_i (f_i · P_i)
where:
  f_i = fraction of tokens dispatched to expert i
  P_i = fraction of router probability allocated to expert i
  N = number of experts
  α = auxiliary loss weight (0.01)
```

#### Router Z-Loss (Zoph et al. ST-MoE)
```
L_z = (1/B) · Σ_x (log Σ_i exp(z_i(x)))²
# Penalizes large logits, improving training stability
```

#### Importance Loss (Shazeer et al.)
```
L_importance = CV(Importance(X))²
where Importance_i(X) = Σ_x G(x)_i  # Sum of gate values for expert i
```

#### Sinkhorn-Knopp Balancing (BASE Layers)
```
# Iteratively normalize rows and columns of score matrix
for iteration in range(sinkhorn_iters):
    S = S / S.sum(dim=1, keepdim=True)  # Normalize rows
    S = S / S.sum(dim=0, keepdim=True)  # Normalize columns
# Result: doubly-stochastic assignment matrix
```

### Expert Revival Strategy
```
1. Monitor per-expert token counts over sliding window
2. If expert_i receives < threshold tokens for K consecutive steps:
   a. Re-initialize expert_i weights from highest-loaded expert + noise
   b. Temporarily boost routing probability to expert_i
   c. Reduce capacity factor for overloaded experts
3. Log revival event for analysis
```

### Dependencies
- Phase 57.1 (ExpertRouter) — routing decisions to balance
- Phase 57.2 (SparseExpertLayer) — expert capacity constraints

### Acceptance Criteria
- [ ] Switch balancing loss with configurable weight
- [ ] Router z-loss implementation (ST-MoE)
- [ ] Importance loss computation
- [ ] Sinkhorn-Knopp doubly-stochastic balancing
- [ ] Dead expert detection and revival
- [ ] Routing collapse detection
- [ ] Adaptive capacity factor scheduling
- [ ] Token dropping with configurable policies
- [ ] Load statistics tracking with EMA smoothing
- [ ] Balance report generation
- [ ] Unit tests with >95% coverage
